File: sdk/javascript/examples/tool-tracing.ts
/**
 * ARKVOID SDK – Tool Execution Tracing Example
 *
 * Demonstrates fine-grained tool call tracing for:
 * - Web search
 * - Code execution
 * - Database queries
 * - External API calls
 * - File operations
 *
 * npm install arkvoid
 */

import { ArkvoidClient } from "arkvoid";
import type { ToolCallRecord, RiskLevel } from "arkvoid";

const arkvoid = new ArkvoidClient({
  apiKey: process.env.ARKVOID_API_KEY!,
  agent: "tool-agent",
  environment: "production",
});

// ─────────────────────────────────────────────
// Tool Registry
// ─────────────────────────────────────────────

interface ToolResult<T = unknown> {
  success: boolean;
  data?: T;
  error?: string;
  latencyMs: number;
}

// Traced tool executor factory
function createTracedTool<TInput extends Record<string, unknown>, TOutput>(
  config: {
    name: string;
    riskLevel: RiskLevel;
    externalSystem?: string;
    execute: (input: TInput) => Promise<TOutput>;
    agentSlug?: string;
  }
) {
  return async (input: TInput): Promise<ToolResult<TOutput>> => {
    const start = Date.now();
    let result: TOutput | undefined;
    let error: string | undefined;

    try {
      result = await config.execute(input);
    } catch (e) {
      error = (e as Error).message;
    }

    const latencyMs = Date.now() - start;

    // Trace each tool call
    await arkvoid.trace({
      action: `tool_call.${config.name}`,
      riskLevel: error ? "high" : config.riskLevel,
      actionType: "tool_call",
      agent: config.agentSlug,
      toolCalls: [
        {
          toolName: config.name,
          callIndex: 0,
          input,
          output: result ? { result } : { error },
          externalSystem: config.externalSystem,
          latencyMs,
          status: error ? "error" : "success",
        },
      ],
      durationMs: latencyMs,
      metadata: {
        success: !error,
        ...(error && { error }),
      },
      tags: ["tool", config.name, error ? "error" : "success"],
    });

    return { success: !error, data: result, error, latencyMs };
  };
}

// ─────────────────────────────────────────────
// Tool Implementations (simulated)
// ─────────────────────────────────────────────

// 1. Web Search
const webSearch = createTracedTool({
  name: "web_search",
  riskLevel: "low",
  externalSystem: "search_api",
  execute: async ({ query }: { query: string }) => {
    await sleep(200 + Math.random() * 300);
    return {
      results: [
        { title: `Result 1 for "${query}"`, url: "https://example.com/1", snippet: "..." },
        { title: `Result 2 for "${query}"`, url: "https://example.com/2", snippet: "..." },
      ],
      totalResults: 1240000,
    };
  },
});

// 2. Code Execution (sandboxed)
const executeCode = createTracedTool({
  name: "code_execution",
  riskLevel: "high",
  externalSystem: "code_sandbox",
  execute: async ({
    code,
    language,
  }: {
    code: string;
    language: string;
  }) => {
    await sleep(500 + Math.random() * 1000);
    // Simulated execution
    if (code.includes("rm -rf") || code.includes("delete")) {
      throw new Error("Dangerous operation detected");
    }
    return {
      stdout: `[Executed ${language} code]\nOutput: 42`,
      stderr: "",
      exitCode: 0,
      executionTime: 0.34,
    };
  },
});

// 3. Database Query
const queryDatabase = createTracedTool({
  name: "database_query",
  riskLevel: "medium",
  externalSystem: "postgres_prod",
  execute: async ({ sql, params }: { sql: string; params: unknown[] }) => {
    await sleep(50 + Math.random() * 200);

    // Detect potential SQL injection pattern
    if (sql.toLowerCase().includes("drop table") || sql.toLowerCase().includes("--")) {
      throw new Error("Suspicious SQL pattern detected");
    }

    return {
      rows: [{ id: 1, name: "Alice" }, { id: 2, name: "Bob" }],
      rowCount: 2,
      duration: 45,
    };
  },
});

// 4. External API Call
const callExternalApi = createTracedTool({
  name: "external_api",
  riskLevel: "medium",
  externalSystem: "payment_gateway",
  execute: async ({
    endpoint,
    payload,
  }: {
    endpoint: string;
    payload: Record<string, unknown>;
  }) => {
    await sleep(100 + Math.random() * 400);
    return {
      statusCode: 200,
      response: { success: true, transactionId: "txn_abc123" },
    };
  },
});

// 5. File Operations
const fileOperation = createTracedTool({
  name: "file_operation",
  riskLevel: "medium",
  externalSystem: "s3_bucket",
  execute: async ({
    operation,
    path,
  }: {
    operation: "read" | "write" | "delete";
    path: string;
  }) => {
    await sleep(80 + Math.random() * 150);

    if (operation === "delete" && path.includes("production")) {
      throw new Error("Cannot delete production files without approval");
    }

    return {
      operation,
      path,
      success: true,
      sizeBytes: operation === "read" ? 4096 : undefined,
    };
  },
});

// ─────────────────────────────────────────────
// Multi-Tool Agent Trace (batch)
// ─────────────────────────────────────────────

async function runResearchWorkflow(topic: string) {
  const toolCalls: ToolCallRecord[] = [];
  const start = Date.now();

  console.log(`\n🔎 Researching: "${topic}"`);

  // Execute multiple tools
  const search1 = await webSearch({ query: `${topic} latest research` });
  const search2 = await webSearch({ query: `${topic} best practices` });
  const dbResult = await queryDatabase({
    sql: "SELECT * FROM knowledge_base WHERE topic = $1 LIMIT 10",
    params: [topic],
  });

  // Aggregate tool calls for a single "research" trace
  if (search1.data) {
    toolCalls.push({
      toolName: "web_search",
      callIndex: 0,
      input: { query: `${topic} latest research` },
      output: { resultCount: (search1.data as { results: unknown[] }).results.length },
      status: "success",
      latencyMs: search1.latencyMs,
    });
  }

  if (search2.data) {
    toolCalls.push({
      toolName: "web_search",
      callIndex: 1,
      input: { query: `${topic} best practices` },
      output: { resultCount: (search2.data as { results: unknown[] }).results.length },
      status: "success",
      latencyMs: search2.latencyMs,
    });
  }

  if (dbResult.data) {
    toolCalls.push({
      toolName: "database_query",
      callIndex: 2,
      input: { table: "knowledge_base", topic },
      output: { rowCount: (dbResult.data as { rowCount: number }).rowCount },
      externalSystem: "postgres_prod",
      status: "success",
      latencyMs: dbResult.latencyMs,
    });
  }

  // Single aggregated trace for the full workflow
  const trace = await arkvoid.trace({
    action: "research_workflow",
    riskLevel: "low",
    durationMs: Date.now() - start,
    toolCalls,
    metadata: {
      topic,
      toolsUsed: toolCalls.map((t) => t.toolName),
      dataSourcesQueried: 3,
    },
    tags: ["research", "multi-tool"],
  });

  console.log(`✅ Research workflow traced: ${trace?.traceId}`);
}

// ─────────────────────────────────────────────
// High-Risk Tool Interception
// ─────────────────────────────────────────────

async function runHighRiskOperation() {
  console.log("\n⚠️  Running high-risk code execution...");

  const result = await executeCode({
    code: "print('Hello, World!')",
    language: "python",
  });

  if (!result.success) {
    // Trace the failure with critical risk
    await arkvoid.trace({
      action: "code_execution_blocked",
      riskLevel: "critical",
      riskScore: 95,
      metadata: {
        reason: result.error,
        blocked: true,
      },
      tags: ["code-exec", "blocked", "security"],
    });
    console.log(`🚫 Blocked: ${result.error}`);
  } else {
    console.log(`✅ Executed safely (${result.latencyMs}ms)`);
  }
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

// ─────────────────────────────────────────────
// Run
// ─────────────────────────────────────────────

(async () => {
  console.log("🔧 ARKVOID Tool Execution Tracing Example\n");

  await runResearchWorkflow("AI agent governance");
  await runHighRiskOperation();

  // Single tool traces
  const apiResult = await callExternalApi({
    endpoint: "/payments/charge",
    payload: { amount: 2999, currency: "usd" },
  });
  console.log(`\n💳 API call: ${apiResult.success ? "success" : apiResult.error}`);

  console.log("\n🎉 All tool traces sent to ARKVOID!");
})();

File: sdk/python/examples/tool_tracing.py
"""
ARKVOID SDK – Tool Execution Tracing (Python)

Demonstrates granular tool tracing using:
- @trace_tool decorator
- ToolCallRecord batching
- Risk-level escalation on tool errors
- Multi-tool workflow aggregation

pip install arkvoid
"""

import os
import random
import time
from typing import Any, Dict, List, Optional

from arkvoid import ArkvoidClient, ToolCallRecord, DataAccessRecord, trace_tool

# ─────────────────────────────────────────────
# Client
# ─────────────────────────────────────────────

arkvoid = ArkvoidClient(
    api_key=os.environ["ARKVOID_API_KEY"],
    agent="tool-agent",
    environment="production",
)


# ─────────────────────────────────────────────
# Tools via @trace_tool decorator
# ─────────────────────────────────────────────

@trace_tool(agent="tool-agent", tool_name="web_search", external_system="serpapi", risk_level="low")
def web_search(query: str) -> Dict[str, Any]:
    """Search the web. Auto-traced via @trace_tool."""
    time.sleep(random.uniform(0.1, 0.4))
    return {
        "results": [
            {"title": f"Result 1 for '{query}'", "url": "https://example.com/1"},
            {"title": f"Result 2 for '{query}'", "url": "https://example.com/2"},
        ],
        "total": 1_240_000,
    }


@trace_tool(agent="tool-agent", tool_name="calculator", risk_level="low")
def calculator(expression: str) -> Dict[str, Any]:
    """Safe math evaluation. Auto-traced via @trace_tool."""
    time.sleep(0.02)
    try:
        result = eval(expression, {"__builtins__": {}}, {})
        return {"result": result, "expression": expression}
    except Exception as e:
        raise ValueError(f"Cannot evaluate: {expression} → {e}") from e


@trace_tool(agent="tool-agent", tool_name="database_query", external_system="postgres_prod", risk_level="medium")
def database_query(sql: str, params: Optional[List[Any]] = None) -> Dict[str, Any]:
    """Query the database. Auto-traced via @trace_tool."""
    time.sleep(random.uniform(0.05, 0.2))
    if any(kw in sql.lower() for kw in ("drop", "delete", "truncate", "--")):
        raise PermissionError("Destructive SQL blocked by governance policy")
    return {
        "rows": [{"id": 1, "value": "alpha"}, {"id": 2, "value": "beta"}],
        "row_count": 2,
        "query_time_ms": 48,
    }


@trace_tool(agent="tool-agent", tool_name="send_email", external_system="sendgrid", risk_level="medium")
def send_email(to: str, subject: str, body: str) -> Dict[str, Any]:
    """Send email via external service. Auto-traced."""
    time.sleep(0.15)
    return {"message_id": f"msg_{int(time.time())}", "status": "sent", "to": to}


@trace_tool(agent="tool-agent", tool_name="file_read", external_system="s3", risk_level="medium")
def read_file(path: str) -> Dict[str, Any]:
    """Read a file from S3. Auto-traced."""
    time.sleep(0.1)
    if "restricted" in path:
        raise PermissionError(f"Access denied: {path} is classified as restricted")
    return {"path": path, "content": f"[Contents of {path}]", "size_bytes": 4096}


# ─────────────────────────────────────────────
# Manual batch tracing (multiple tools → 1 trace)
# ─────────────────────────────────────────────

def run_research_workflow(topic: str) -> Dict[str, Any]:
    """Execute a multi-tool research workflow and record as a single trace."""
    tool_calls_recorded: List[ToolCallRecord] = []
    workflow_start = time.time()

    print(f"\n🔍 Research workflow: '{topic}'")

    # Tool 1: web search
    t = time.time()
    search1 = web_search(f"{topic} overview")
    tool_calls_recorded.append(ToolCallRecord(
        tool_name="web_search", call_index=0,
        input={"query": f"{topic} overview"},
        output={"result_count": len(search1.get("results", []))},
        latency_ms=int((time.time() - t) * 1000), status="success",
    ))

    # Tool 2: second search
    t = time.time()
    search2 = web_search(f"{topic} best practices 2025")
    tool_calls_recorded.append(ToolCallRecord(
        tool_name="web_search", call_index=1,
        input={"query": f"{topic} best practices 2025"},
        output={"result_count": len(search2.get("results", []))},
        latency_ms=int((time.time() - t) * 1000), status="success",
    ))

    # Tool 3: DB lookup
    t = time.time()
    db_result = database_query("SELECT * FROM knowledge_base WHERE topic = %s LIMIT 5", [topic])
    tool_calls_recorded.append(ToolCallRecord(
        tool_name="database_query", call_index=2,
        input={"table": "knowledge_base", "topic": topic},
        output={"row_count": db_result.get("row_count", 0)},
        external_system="postgres_prod",
        latency_ms=int((time.time() - t) * 1000), status="success",
    ))

    # Single aggregated trace for the whole workflow
    trace_result = arkvoid.trace(
        action="research_workflow",
        risk_level="low",
        duration_ms=int((time.time() - workflow_start) * 1000),
        tool_calls=tool_calls_recorded,
        metadata={
            "topic": topic,
            "tools_used": list({tc.tool_name for tc in tool_calls_recorded}),
            "total_tool_calls": len(tool_calls_recorded),
        },
        tags=["research", "multi-tool"],
    )

    print(f"  ✅ Workflow trace: {trace_result.trace_id if trace_result else '–'}")
    return {"search": search1, "db": db_result}


# ─────────────────────────────────────────────
# High-risk blocked operation
# ─────────────────────────────────────────────

def run_risky_operation():
    print("\n⚠️  Attempting risky DB operation...")
    start = time.time()
    try:
        database_query("DROP TABLE users --")
    except PermissionError as e:
        # Log the block as a critical trace
        arkvoid.trace(
            action="dangerous_query_blocked",
            risk_level="critical",
            risk_score=98,
            duration_ms=int((time.time() - start) * 1000),
            metadata={
                "reason": str(e),
                "blocked_by": "arkvoid_governance",
            },
            tags=["security", "blocked", "sql-injection"],
        )
        print(f"  🚫 Blocked & traced: {e}")


# ─────────────────────────────────────────────
# Parallel tool execution
# ─────────────────────────────────────────────

def run_parallel_tools():
    import threading

    print("\n⚡ Parallel tool execution...")
    results: Dict[str, Any] = {}

    def run_search(q: str):
        results[f"search_{q[:10]}"] = web_search(q)

    threads = [
        threading.Thread(target=run_search, args=(f"AI governance topic {i}",))
        for i in range(3)
    ]
    [t.start() for t in threads]
    [t.join() for t in threads]

    arkvoid.trace(
        action="parallel_search",
        risk_level="low",
        metadata={"parallel_count": 3, "results_count": len(results)},
        tags=["parallel", "search"],
    )
    print(f"  ✅ {len(results)} parallel searches traced")


# ─────────────────────────────────────────────
# Run
# ─────────────────────────────────────────────

if __name__ == "__main__":
    print("🔧 ARKVOID Tool Execution Tracing (Python)\n")

    # Multi-tool workflow
    run_research_workflow("AI agent governance")

    # High-risk block
    run_risky_operation()

    # Parallel tools
    run_parallel_tools()

    # Single tool with data access
    print("\n📨 Sending email...")
    result = send_email(
        to="analyst@company.com",
        subject="Research Report Ready",
        body="Your AI governance report has been generated.",
    )
    print(f"  ✅ Email sent: {result.get('message_id')}")

    print("\n🎉 All tool traces sent to ARKVOID!")
